File: maxmoney_iterative.c
#include <stdio.h>
// autotype using https://www.askpython.com/python/examples/auto-type-text-using-python
int fact(int n)
{   int fact=1,i=1;
    do
    {
        fact*=i;
        i++;
    }
    while (i<=n);
    return(fact);
}
int  ** ncr(int n,int r)
{
    if(r==0)
    {
        int p=-1;
        int *q=&p;
        int **r=&q;
        return r;
    }
    else if(r==n)
    {
        int * lista;
        for (int i=1;i<n+1;i++)
        {
            int *p=&i;
            lista[i]=*p;
        }
        int **listb=&lista;
        return listb;
    }
    else
    {
        int ** templist;
        for(int i=1;i<n-r+2;i++)
        {
            int ** temp= ncr(n-1,r-1);
            int nchooser=fact(n-1)/(fact(r-1)*fact(n-r));
            for(int j=0;j<nchooser;j++)
            {
                int flag=1,count=0;
                for(int k=0;k<r-1;k++)
                {
                    (** (temp + (j*(r-1) + k)))++;
                    if ((** (temp + (j*(r-1) + k)))>n)
                    flag=0;
                }
                if (flag)
                {
                    for(int k=0;k<r-1;k++)
                    {
                        **(templist+count)=**(temp + (j*(r-1) + k));
                        count++;
                    }
                }
            }
        }
        return templist;
    }
}

int main()
{
    int n=4,r=2;
    printf("heklo");
}

File: maxmoney_iterative.py

def ncr(n,r):
    if r==0:
        return([[]])
    elif r==n:
        lista=[]
        for i in range(1,n+1):
            lista+=[i]
        return([lista])
    else:
        templist=[]
        for i in range(1,n-r+2):
            for j in (ncr(n-1,r-1)):
                flag=1
                for k in range(len(j)):
                    j[k]+=i
                    if j[k]>n:
                        flag=0
                if (flag): templist+=[[i]+j]
        return(templist)
houses=[3,9,4,2,8]
n=len(houses)
possible=[]
for i in range(1,n+1):
    g=ncr(n,i)
    for a in g:
        for q in range(len(a)):
            a[q]-=1

    for a in g:
        valid=1
        for j in range(len(a)-1):
            if a[j]+1==a[j+1]:
                valid=0
        if valid: possible+=[a]

maxloot=0
for i in possible:
    loot=0
    for j in i:
        loot+=houses[j]
    if loot>maxloot: maxloot=loot
print(maxloot)